clang/lib/Sema/SemaExprCXX.cpp

```diff
@@ -5637,6 +5638,27 @@ static bool EvaluateUnaryTypeTrait(Sema &Self, TypeTrait UTT,
     return false;
   case UTT_IsTriviallyEqualityComparable:
     return isTriviallyEqualityComparableType(Self, T, KeyLoc);
+  case UTT_IsImplicitLifetime: {
+    DiagnoseVLAInCXXTypeTrait(Self, TInfo,
+                              tok::kw___builtin_is_implicit_lifetime);
+    QualType UnqualT = T->getCanonicalTypeUnqualified();
+    if (UnqualT->isScalarType())
+      return true;
+    if (UnqualT->isArrayType())
+      return true;
+
+    const CXXRecordDecl *RD = UnqualT->getAsCXXRecordDecl();
+    if (!RD)
+      return false;
+    if (UnqualT->isAggregateType())
+      if (!RD->getDestructor()->isUserProvided())
+        return true;
+    if (RD->hasTrivialDestructor())
```

Comment:
You also need to check that it's not deleted too.

Currently the first two assertions fail here:

```c++
struct X { ~X() = delete; };
struct Y { X x; ~Y() = default; };
struct Z { X x; };

static_assert(!__builtin_is_implicit_lifetime(X));
static_assert(!__builtin_is_implicit_lifetime(Y));
static_assert( __builtin_is_implicit_lifetime(Z));
```

Reply:
All three types in your examples are implicit lifetime class types, because they are aggregates and does not have a user-provided destructor.

Reply:
@Endilll That's correct for `Z` but `X` and `Y` are not aggregates because they have user-declared constructors, so they need a non-deleted destructor

Reply:
> but X and Y are not aggregates because they have user-declared constructors

I'm sorry, but they only thing I see there are user-declared _destructors_, not constructors.